hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -19,29 +38,232 @@ impl TraceDecoder for YkPTTraceDecoder {
+            // We know where we are in the main object binary, so there's a chance that there's a
+            // blockmap entry for this location.
+            let ent = Self::lookup_blockmap_entry(b_off.to_owned());
+            if ent.is_none() {
+                // In the absence of blockmap info, we cannot know the successor block.
+                self.cur_loc = ObjLoc::OtherObjOrUnknown;
+                return Ok(Some(Block::unknown()));
+            }
+            let ent = ent.unwrap();
+
+            // If there are calls in the block that come *after* the current position in the block,
+            // then we will need to follow those before we look at the successor info.
+            if let Some(call_info) = ent
+                .value
+                .call_offs()
+                .iter()
+                .find(|c| c.callsite_off() > b_off)
+            {
+                let target = call_info.target_off();
+                if let Some(target_off) = target {
+                    self.cur_loc = ObjLoc::MainObj(target_off);
+                    return Ok(Some(self.lookup_block_from_main_bin_offset(target_off)));
+                } else {
+                    // Call target isn't known statically.
+                    self.cur_loc = ObjLoc::OtherObjOrUnknown;
+                    return Ok(Some(Block::unknown()));
+                }
+            }
+
+            // If we get here, there were no further calls to follow in the block, so we consult
+            // the static successor information.
+            match ent.value.successor() {
+                SuccessorKind::Unconditional { target } => {
+                    if let Some(target_off) = target {
+                        self.cur_loc = ObjLoc::MainObj(*target_off);
+                        return Ok(Some(self.lookup_block_from_main_bin_offset(*target_off)));
+                    } else {
+                        // Divergent control flow.
+                        todo!();
+                    }
+                }
+                SuccessorKind::Conditional {
+                    taken_target,
+                    not_taken_target,
+                } => {
+                    // If we don't have any TNT choices buffered, get more.
+                    if self.tnts.is_empty() {
+                        if self.seek_tnt()? == false {
+                            // The packet stream was exhausted.
+                            return Ok(None);
+                        }
+                    }
+                    // Find where to go next based on whether the trace took this branch or not.
+                    let target_off = if self.tnts.pop_front().unwrap() {
+                        *taken_target
+                    } else {
+                        if let Some(ntt) = not_taken_target {
+                            *ntt
+                        } else {
+                            // Divergent control flow.
+                            panic!();
```

Comment:
`panic` or `todo`?

Reply:
That should be a `todo!()`. My bad.

Reply:
Fixed in e1c20f15a7781308efdffb90052da762bc121a8e